feat: TLS certificate rotation (#678)
Azure/aks-engine: docs/design/cert-rotation.md

# Cluster Certificate Rotation Feature Design

> Design doc on the certificate rotation feature for AKS Engine clusters.

Reviewer: ```suggestion
Design document for the certificate rotation feature in AKS Engine clusters.
```

## Problem

Until PR [#396](https://github.com/Azure/aks-engine/pull/396), the default expiration of the certs generated by AKS Engine was 2 years. Clusters that were deployed with an AKS Engine version prior to that PR will soon reach expiration. We currently do not have an outlined process for rotating certificates in an existing cluster.

## Goals

Provide a process that is easy to follow for rotating CA, etcd, kubelet, kubeconfig and apiserver certificates in a cluster built with AKS Engine.

## Non-Goals

> - Providing a certificate rotation tool for AKS clusters

Reviewer: Full stops here, please.

Author: you mean periods?

Author: nevermind, it became clear in the following comments :)

Reviewer: Yeah. I try to refrain from AmE as much as possible :P.

- Providing a tool for etcd backup
- Rotating certificates in an existing cluster with no downtime

## Alternative implementations considered

### Using upgrade to update the certificates

What: Implementing a new "certificate-rotation" command that runs upgrade under the hood. This command would perform a same k8s version upgrade on the cluster with newly generated certificates replacing the old ones in the apimodel.

Pros:

> - No reboot required after the fact

Reviewer: ```suggestion
- No reboot required after the fact.
```

Cons:

> - Requires re-provisionning all the VMs, causing unnecessary time overhead

Reviewer: ```suggestion
- Requires a re-provisioning of all the VMs, causing an unnecessary time overhead.
```

> - Current implementation does not support this: the newly built master nodes would not be able to join the etcd cluster, thus failing CSE validation

Reviewer: ```suggestion
- Current implementation does not support this: the newly built master nodes would not be able to join the etcd cluster, thus failing CSE validation.
```

### Bash scripts

What: Provide users with a bash script that they can run to rotate certificates on their cluster.

Pros:

> - Easier to implement

Reviewer: Full stops for these phrases please

- More lightweight
- Performs in-place cert update

Cons:

- Requires more manual steps from the user. For example, they would have to set environemnent variables for their SSH key, KUBECONFIG, _output directory, etc.
- Not Windows friendly: would require running the script in a container, unless we provide a Powershell equivalent

### New aks-engine binary command

What: New `aks-engine rotate-certs` command that uses a combination of Kubernetes client-go and SSH commands to access the cluster and rotate certificates.

Pros:

> - Better UX

Reviewer: I think you'd know what I'm gonna by now :P

Author: haha
